packages/vite/src/node/optimizer/index.ts

```diff
@@ -587,50 +610,77 @@ export function runOptimizeDeps(
-
-            // Get a list of old files in the deps directory to delete the stale ones
-            const oldFilesPaths: string[] = []
-            if (!fs.existsSync(depsCacheDir)) {
-              fs.mkdirSync(depsCacheDir, { recursive: true })
-            } else {
-              oldFilesPaths.push(
-                ...(await fsp.readdir(depsCacheDir)).map((f) =>
-                  path.join(depsCacheDir, f),
-                ),
-              )
-            }
+        async function commitFiles() {
+          // Write this run of pre-bundled dependencies to the deps cache
 
-            const newFilesPaths = new Set<string>()
-            const files: Promise<void>[] = []
-            const write = (filePath: string, content: string) => {
-              newFilesPaths.add(filePath)
-              files.push(fsp.writeFile(filePath, content))
-            }
+          // Keep the output files in memory while we write them to disk in the
+          // background. These files are going to be sent right away to the browser
+          optimizedDepsCache.set(
+            metadata,
+            new Map(
+              result.outputFiles!.map((f) => [normalizePath(f.path), f.text]),
+            ),
+          )
+
+          // Get a list of old files in the deps directory to delete the stale ones
+          const oldFilesPaths: string[] = []
+          if (!fs.existsSync(depsCacheDir)) {
+            fs.mkdirSync(depsCacheDir, { recursive: true })
+          } else {
+            oldFilesPaths.push(
+              ...(await fsp.readdir(depsCacheDir)).map((f) =>
+                path.join(depsCacheDir, f),
+              ),
+            )
+          }
+
+          const writingFilePath = path.resolve(depsCacheDir, '_writing')
+          await fsp.writeFile(writingFilePath, '')
 
+          const newFilesPaths = new Set<string>()
+          const files: Promise<void>[] = []
+          const write = (filePath: string, content: string) => {
+            newFilesPaths.add(filePath)
+            files.push(fsp.writeFile(filePath, content))
+          }
+
+          path.join(depsCacheDir, '_metadata.json'),
             // a hint for Node.js
             // all files in the cache directory should be recognized as ES modules
             write(
               path.resolve(depsCacheDir, 'package.json'),
               '{\n  "type": "module"\n}\n',
             )
 
-            write(
-              path.join(depsCacheDir, '_metadata.json'),
-              stringifyDepsOptimizerMetadata(metadata, depsCacheDir),
-            )
+          write(
+            path.join(depsCacheDir, '_metadata.json'),
+            stringifyDepsOptimizerMetadata(metadata, depsCacheDir),
+          )
 
-            for (const outputFile of result.outputFiles!)
-              write(outputFile.path, outputFile.text)
+          for (const outputFile of result.outputFiles!)
+            write(outputFile.path, outputFile.text)
```

Comment:
```suggestion
            write(outputFile.path, outputFile.contents)
```

Realized a small optimization here, maybe we can write the file with the buffer directly. The `.text` is a getter that's derived from the buffer.